tsk/util/Bitlocker/BitlockerParser.cpp

```diff
@@ -0,0 +1,1553 @@
+        // recorded clear key was incorrect/corrupted. We don't want to tell the user that the
+        // password they probably didn't even enter is incorrect.
+        writeError("BitlockerParser::parseVMKEntry: Failed to decrypt VMK using the supplied clear key");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    // Make sure the value is of type Key
+    if (keyEntry->getValueType() != BITLOCKER_METADATA_VALUE_TYPE::KEY) {
+        writeError("BitlockerParser::parseVMKEntry: keyEntry does not have value of type KEY ("
+            + convertMetadataValueTypeToString(keyEntry->getValueType()) + ")");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    // Save the decrypted VMK and what method we used to decrypt it.
+    *vmkEntry = keyEntry;
+    m_protectionTypeUsed = protectionType;
+
+    return BITLOCKER_STATUS::SUCCESS;
+}
+
+/**
+* Use the decrypted volume master key (VMK) entry to get the full volume encryption key (FVEK).
+* We should have set m_decryptedVmkEntry prior to calling this method.
+* 
+* @return SUCCESS on success, GENERAL_ERROR otherwise.
+*/
+BITLOCKER_STATUS BitlockerParser::getFullVolumeEncryptionKey() {
+
+    // Sanity check
+    if (m_decryptedVmkEntry == nullptr) {
+        writeError("BitlockerParser::getFullVolumeEncryptionKey: VMK is not set");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    // Find the FVEK entry
+    list<MetadataEntry*> fvekEntries;
+    getMetadataEntries(m_metadataEntries, BITLOCKER_METADATA_ENTRY_TYPE::FULL_VOLUME_ENCRYPTION_KEY, 
+        BITLOCKER_METADATA_VALUE_TYPE::AES_CCM_ENCRYPTED_KEY, fvekEntries);
+    if (fvekEntries.empty()) {
+        writeError("BitlockerParser::getFullVolumeEncryptionKey: Could not find FVEK metatdata entry");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    MetadataValueAesCcmEncryptedKey* aesCcmKey = dynamic_cast<MetadataValueAesCcmEncryptedKey*>((fvekEntries.front())->getValue());
+    if (aesCcmKey == nullptr) {
+        writeError("BitlockerParser::getFullVolumeEncryptionKey: Error casting MetadataValueAesCcmEncryptedKey");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    // Decrypt it using the VMK
+    // First get the decrypted key out of the decrypted VMK entry
+    uint8_t* keyBytes = nullptr;
+    size_t keyLen = 0;
+    if (BITLOCKER_STATUS::SUCCESS != getKeyData(m_decryptedVmkEntry, &keyBytes, keyLen)) {
+        writeError("BitlockerParser::getFullVolumeEncryptionKey: Error loading keys");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    // Then use that key to decrypt the FVEK entry, which should produce a MetadataEntry of type KEY.
+    // This includes testing a 16-byte message authentication code to verify that
+    // the decrypted key is correct.
+    MetadataEntry* keyEntry = nullptr;
+    BITLOCKER_STATUS ret = aesCcmKey->decrypt(keyBytes, keyLen, &keyEntry);
+    if (ret != BITLOCKER_STATUS::SUCCESS || keyEntry == nullptr) {
```

Comment:
other calls to decrypt() only check the status we should be consistent if we should be checking for keyEntry being nullptr